From f4b4fdaeff95125b31d5b9190b978649d1eceee4 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 28 Nov 2014 16:18:36 -0500
Subject: [md] dm-cache: only use overwrite optimisation for promotion when in writeback mode

Message-id: <1417191518-16689-8-git-send-email-snitzer@redhat.com>
Patchwork-id: 101009
O-Subject: [RHEL7.1 PATCH 7/9] dm cache: only use overwrite optimisation for promotion when in writeback mode
Bugzilla: 1165050
RH-Acked-by: Joe Thornber <thornber@redhat.com>
RH-Acked-by: Heinz Mauelshagen <heinzm@redhat.com>

BZ: 1165050

Upstream linux-dm.git commit fa607bcadcb717601e5ceaa2293fefdd14774165
Author: Joe Thornber <ejt@redhat.com>
Date:   Thu Nov 27 12:21:08 2014 +0000

    dm cache: only use overwrite optimisation for promotion when in writeback mode

    Overwrite causes the cache block and origin blocks to diverge, which
    is only allowed in writeback mode.

    Signed-off-by: Joe Thornber <ejt@redhat.com>
    Signed-off-by: Mike Snitzer <snitzer@redhat.com>
    Cc: stable@vger.kernel.org

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm-cache-target.c b/drivers/md/dm-cache-target.c
index 856d0c0..ea53d76 100644
--- a/drivers/md/dm-cache-target.c
+++ b/drivers/md/dm-cache-target.c
@@ -1136,7 +1136,8 @@ static void issue_copy_or_discard(struct dm_cache_migration *mg)
 
   avoid = is_discarded_oblock(cache, mg->new_oblock);
 
-  if (!avoid && bio_writes_complete_block(cache, bio)) {
+  if (writeback_mode(&cache->features) &&
+      !avoid && bio_writes_complete_block(cache, bio)) {
    issue_overwrite(mg, bio);
    return;
   }
-- 
1.7.1